src/components/structures/FavouriteMessagesView/FavouriteMessagesView.tsx

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React, { FC, useCallback, useContext, useEffect, useState } from "react";
+import { MatrixClient, MatrixEvent, RelationType } from "matrix-js-sdk/src/matrix";
+import { logger } from "matrix-js-sdk/src/logger";
+
+import MatrixClientContext from "../../../contexts/MatrixClientContext";
+import useFavouriteMessages from "../../../hooks/useFavouriteMessages";
+import ResizeNotifier from "../../../utils/ResizeNotifier";
+import FavouriteMessagesPanel from "./FavouriteMessagesPanel";
+import { FavouriteMessagesStore, FavouriteStorage } from "../../../stores/FavouriteMessagesStore";
+
+interface IProps {
+    resizeNotifier?: ResizeNotifier;
+
+    // Provide this to use a different store for favourite messages
+    // e.g. in tests. If not supplied, we use the global default.
+    favouriteMessagesStore?: FavouriteMessagesStore;
+}
+
+const FavouriteMessagesView: FC<IProps> = ({ resizeNotifier, favouriteMessagesStore }: IProps) => {
+    const matrixClient = useContext<MatrixClient>(MatrixClientContext);
+    const { getFavouriteMessages, registerFavouritesChangedListener } = useFavouriteMessages(favouriteMessagesStore);
+    const [searchQuery, setSearchQuery] = useState("");
+    const [favouriteMessageEvents, setFavouriteMessageEvents] = useState<MatrixEvent[] | null>(null);
+
+    const recalcEvents = useCallback(async () => {
+        const faves = getFavouriteMessages();
+        const newEvents = await calcEvents(searchQuery, faves, matrixClient);
+        setFavouriteMessageEvents(newEvents);
+    }, [searchQuery, matrixClient, getFavouriteMessages]);
+
+    // Because finding events is async, we do it in useEffect, not useState.
+    useEffect(() => {
+        recalcEvents();
+    }, [searchQuery, recalcEvents]);
+
+    registerFavouritesChangedListener(() => recalcEvents());
```

Comment:
At every render, it will call the `registerFavouritesChangedListener`  and a listener

Reply:
Do you have an alternative suggestion?

Reply:
We can add the listener in a `useEffect`:

```
useEffect(() => {
     const handler = () => recalcEvents();
     registerFavouritesChangedListener(handler);
     return () => {
         // here we would like to unregister the handler
         unregisterFavouritesChangedListener(handler);
     }
}, [registerFavouritesChangedListener, recalcEvents]
```

If we want to avoid to repeat this behaviour every time we use the `useFavourites` hook. We can pass the handler as argument of the hook and make all the register/unregister in it.